### ISSUE TITLE
401 error on fetching playlist iOS

### ISSUE THREAD

**Issue by MusicMonster20** [soundcloud/api#230]
Hi Everyone,
We are implementing sound cloud functionality in our iOS app and trying to fetch the user playlist. 

We are using the following API to get access token: 
curl -X POST "https://api.soundcloud.com/oauth2/token" -H  "accept: application/json; charset=utf-8" -H  "Content-Type: application/x-www-form-urlencoded" -d "grant_type=client_credentials&client_id=**********************&client_secret=**************************&code=3-292653-1184522413-9z87Q1rA8mpxH&redirect_uri=https%3A%2F%2Fmywebsite%2Fauth%2Fsoundcloud&refresh_token="

And after getting access token successfully, we are hitting following API to get the playlist 
curl -X GET "https://api.soundcloud.com/me/playlists?show_tracks=true&linked_partitioning=true&limit=2" -H  "accept: application/json; charset=utf-8" -H  "Authorization: OAuth 2-292653--ATesTnYIoWRn4pClxLh67BD

We are getting 401-Unauthorized error every time. What we are missing here ? Kindly advise.  Thank you in advance.

**Comment by dpreussler** [comment]
Hi there, is the user you are trying to get the playlist from the same that you authenticated?

**Comment by mgoodfellow** [comment]
It doesn't matter what user they are authenticated as because they are requesting the `/me` route - so this would serve up the playlists for that user context.

Good spot though, they aren't actually authenticated as a user at all! They are using a `client_credentials` grant which is a server token with no user context and therefore it cannot access any of the `/me/...` resources.

```Client Credentials Token Exchange Flow
With applications, such as CLIs, or pure back-end services you would authenticate the application itself rather than a user. For instance, you might want to build an artist's website where you only need an access to their tracks, playlists, or user information. There is no need to go through the connect flow, as SoundCloud API provides the Client Credentials Flow for these purposes. You pass along the client_id and client_secret you have acquired at registration to authenticate and get a token.
```

@MusicMonster20 check out: https://developers.soundcloud.com/docs/api/guide#authentication

TL;DR you should be using `authorization_code` grant flow because you are providing a `code` parameter.

**Comment by dpreussler** [comment]
good catch, thanks @mgoodfellow

**Comment by MusicMonster20** [comment]
Hi, kindly check this: https://prnt.sc/Qqk_cCq9Iyde there is an error of invalid_grant.

**Comment by mgoodfellow** [comment]
If I had to guess, which is all I can do - I would say you have not sent your redirect_uri exactly as it is configured for your app.

The example you gave in your first comment doesn't look like the right form. Why would it end in `refresh_token=`?

**Comment by MusicMonster20** [comment]
Hi, so you mean we do not require refresh_token= at the end ? 
So it will look like: "grant_type=authorization_code&client_id=**********************&client_secret=**************************&code=3-292653-1184522413-9z87Q1rA8mpxH&redirect_uri=***********://soundcloud/redirect" ?

**Comment by mgoodfellow** [comment]
Go to https://soundcloud.com/you/apps - copy what you have listed as the Redirect URI and put as the `redirect_uri`

**Comment by davidjohnson85** [comment]
The 401 Unauthorized error suggests that there is an issue with the authentication or authorization process. Here are a few things to check:

1. Ensure the access token is included correctly: Double-check that the access token is included in the request headers with the correct format. In your provided cURL command, it seems that the access token is included in the Authorization header as "OAuth 2-292653--ATesTnYIoWRn4pClxLh67BD". Verify that the access token is valid and has the necessary permissions to access the user's playlists.

2. Verify the API endpoint: Confirm that the API endpoint you are using to fetch the playlists is correct. In your example, the endpoint is "https://api.soundcloud.com/me/playlists". Ensure that this endpoint is the correct one for fetching a user's playlists. Additionally, make sure the API version and any required parameters are included correctly.

3. Check the OAuth scopes: Make sure that the access token you are obtaining has the necessary scopes to access the user's playlists. In your case, the access token is obtained using the "client_credentials" grant type, which may have limitations on the available scopes. Verify that you are requesting the appropriate scope(s) during the authorization process.

4. Test the API with a different tool: Use a tool like cURL or an API testing tool (e.g., Postman) to make the same API request and see if you encounter the same issue. This can help verify if the problem lies with the code or configuration in your iOS app.

5. Review SoundCloud's API documentation: Refer to SoundCloud's API documentation for any specific requirements or considerations when fetching user playlists. Pay attention to any changes in the API endpoints, authentication methods, or rate limiting policies.

If the issue persists after checking these points, it may be necessary to contact SoundCloud's developer support for further assistance. They can provide insights into any potential issues with your authentication or offer guidance on resolving the 401 Unauthorized error.

**Regards;
_David Johnson_.**
